File: timesheet-example/workflows-java/src/main/java/com/example/flow/SignHandler.java
package com.example.flow;

import co.paralleluniverse.fibers.Suspendable;
import com.example.service.SalaryRateOracle;
import kotlin.Unit;

import net.corda.core.crypto.TransactionSignature;
import net.corda.core.flows.FlowException;
import net.corda.core.flows.FlowLogic;
import net.corda.core.flows.FlowSession;
import net.corda.core.flows.InitiatedBy;
import net.corda.core.transactions.FilteredTransaction;
import net.corda.core.utilities.ProgressTracker;

@InitiatedBy(IssueInvoiceFlow.Acceptor.class)
public class SignHandler extends FlowLogic<Unit> {
    private ProgressTracker.Step RECEIVING = new ProgressTracker.Step("Receiving sign request");
    private ProgressTracker.Step SIGNING = new ProgressTracker.Step("Signing filtered transaction.");
    private ProgressTracker.Step SENDING = new ProgressTracker.Step("Sending sign response.");
    private FlowSession session;

    public SignHandler(FlowSession session) {
        this.session = session;
    }

    private final ProgressTracker progressTracker = new ProgressTracker(
            RECEIVING,
            SIGNING,
            SENDING
    );

    @Suspendable
    @Override
    public Unit call() throws FlowException {
        progressTracker.setCurrentStep(RECEIVING);
        FilteredTransaction request = session.receive(FilteredTransaction.class).unwrap( it -> it);

        progressTracker.setCurrentStep(SIGNING);
        try{
            TransactionSignature response = getServiceHub().cordaService(SalaryRateOracle.class).sign(request);
            progressTracker.setCurrentStep(SENDING);
            session.send(response);
        } catch(Exception e ){
            throw new FlowException(e);
        };
        return null;
    }
}

File: timesheet-example/workflows-kotlin/src/main/kotlin/com/example/flow/SignHandler.kt
package com.example.flow

import co.paralleluniverse.fibers.Suspendable
import net.corda.core.flows.FlowException
import net.corda.core.flows.FlowLogic
import net.corda.core.flows.FlowSession
import net.corda.core.flows.InitiatedBy
import net.corda.core.transactions.FilteredTransaction
import net.corda.core.utilities.ProgressTracker
import net.corda.core.utilities.unwrap
import com.example.service.SalaryRateOracle

// The oracle flow to sign salary rate lookups
@InitiatedBy(IssueInvoiceFlow.Acceptor::class)
class SignHandler(val session: FlowSession) : FlowLogic<Unit>() {
    companion object {
        object RECEIVING : ProgressTracker.Step("Receiving sign request.")
        object SIGNING : ProgressTracker.Step("Signing filtered transaction.")
        object SENDING : ProgressTracker.Step("Sending sign response.")
    }

    override val progressTracker = ProgressTracker(RECEIVING, SIGNING, SENDING)

    @Suspendable
    override fun call() {
        progressTracker.currentStep = RECEIVING
        val request = session.receive<FilteredTransaction>().unwrap { it }

        progressTracker.currentStep = SIGNING
        val response = try {
            serviceHub.cordaService(SalaryRateOracle::class.java).sign(request)
        } catch (e: Exception) {
            throw FlowException(e)
        }

        progressTracker.currentStep = SENDING
        session.send(response)
    }
}
